crates/scroll/evm/src/execute.rs

```diff
@@ -0,0 +1,545 @@
+use reth_evm::{
+    execute::{
+        BlockExecutionError, BlockExecutionStrategy, BlockExecutionStrategyFactory,
+        BlockValidationError, ExecuteOutput, ProviderError,
+    },
+    ConfigureEvm, ConfigureEvmEnv,
+};
+use reth_primitives::{
+    gas_spent_by_transactions, BlockWithSenders, GotExpected, InvalidTransactionError, Receipt,
+};
+use reth_revm::primitives::{CfgEnvWithHandlerCfg, U256};
+use reth_scroll_chainspec::{ScrollChainConfig, ScrollChainSpec};
+use reth_scroll_consensus::apply_curie_hard_fork;
+use reth_scroll_execution::FinalizeExecution;
+use reth_scroll_forks::ScrollHardfork;
+use revm::{
+    db::BundleState,
+    primitives::{bytes::BytesMut, BlockEnv, EnvWithHandlerCfg, ResultAndState},
+    Database, DatabaseCommit, State,
+};
+use std::{
+    fmt::{Debug, Display},
+    sync::Arc,
+};
+
+/// The Scroll block execution strategy.
+#[derive(Debug)]
+pub struct ScrollExecutionStrategy<DB, EvmConfig> {
+    /// Chain specification.
+    chain_spec: Arc<ScrollChainSpec>,
+    /// Evm configuration.
+    evm_config: EvmConfig,
+    /// Current state for the execution.
+    state: State<DB>,
+}
+
+impl<DB, EvmConfig> ScrollExecutionStrategy<DB, EvmConfig> {
+    /// Returns an instance of [`ScrollExecutionStrategy`].
+    pub const fn new(
+        chain_spec: Arc<ScrollChainSpec>,
+        evm_config: EvmConfig,
+        state: State<DB>,
+    ) -> Self {
+        Self { chain_spec, evm_config, state }
+    }
+}
+
+impl<DB, EvmConfig> ScrollExecutionStrategy<DB, EvmConfig>
+where
+    EvmConfig: ConfigureEvmEnv<Header = Header>,
+{
+    /// Configures a new evm configuration and block environment for the given block.
+    ///
+    /// # Caution
+    ///
+    /// This does not initialize the tx environment.
+    fn evm_env_for_block(&self, header: &Header, total_difficulty: U256) -> EnvWithHandlerCfg {
+        let mut cfg = CfgEnvWithHandlerCfg::new(Default::default(), Default::default());
+        let mut block_env = BlockEnv::default();
+        self.evm_config.fill_cfg_and_block_env(&mut cfg, &mut block_env, header, total_difficulty);
+
+        EnvWithHandlerCfg::new_with_cfg_env(cfg, block_env, Default::default())
+    }
+}
+
+impl<DB, EvmConfig> BlockExecutionStrategy<DB> for ScrollExecutionStrategy<DB, EvmConfig>
+where
+    DB: Database<Error: Into<ProviderError> + Display>,
+    State<DB>: FinalizeExecution<Output = BundleState>,
+    EvmConfig: ConfigureEvm<Header = Header>,
+{
+    type Error = BlockExecutionError;
+
+    fn apply_pre_execution_changes(
+        &mut self,
+        block: &BlockWithSenders,
+        _total_difficulty: U256,
+    ) -> Result<(), Self::Error> {
+        if self.chain_spec.fork(ScrollHardfork::Curie).transitions_at_block(block.number) {
+            if let Err(err) = apply_curie_hard_fork(&mut self.state) {
+                tracing::debug!(%err, "failed to apply curie hardfork");
+                return Err(ForkError::Curie.into());
+            };
+        }
```

Comment:
Wondering if there is any gas metering for this in l2geth?

Reply:
I don't think so, the state just gets updated without any gas being accounted in the block.
https://github.com/scroll-tech/go-ethereum/blob/develop/consensus/misc/curie.go#L12-L23